### PR TITLE
feat: support CSS custom property theme maps behind feature flag

### LINKED ISSUE #4033: Create CSS Custom Properties feature flag
We should create a feature flag for enabling css custom properties, probably `enable-css-custom-properties` following the trend of Chrome feature flags (and could set a precedent in the future).

When this flag is enabled, we should redefine our tokens to correspond to the CSS Custom Property equivalent of their values, meaning:

```scss
// Before
$interactive-01: $blue-60
// After
$interactive-01: var(--cds-interactive-01, $blue-60);
```

### PR DESCRIPTION
**Pull request by emyarod**
Closes #4033

refs #4065 #4082 #4094

This PR enables toggling the theme map output behind the `enable-css-custom-properties` feature flag. When the flag is enabled, the theme maps will be generated with custom property map values

#### Changelog

**New**

- `emit-css-custom-properties` feature flag
- optional CSS custom property theme maps

**Changed**

- `@carbon/themes` build task can now generate CSS custom properties and assign them to Sass variables

**Removed**

- legacy theme switcher

#### Testing / Reviewing

Ensure that the correct prefixed CSS custom properties are being emitted (you can also verify this by examining vanilla storybook `html` in dev tools or the `html` within the storybook `iframe` in the React storybook).

Ensure that the theme switchers function as expected

### CODE REVIEW

File: packages/components/demo/scss/demo.scss [hunk 1]
```diff
@@ -20,16 +20,18 @@
 //     interactive-01: pink,
```
Comment (emyarod) [inline packages/components/demo/scss/demo.scss:20]: most likely we can change or remove these comments as they will be outdated

File: packages/components/src/globals/scss/_feature-flags.scss [hunk 2]
```diff
@@ -32,6 +32,7 @@ $default-feature-flags: (
   grid: true,
   grid-columns-16: false,
   grid--fallback: false,
+  enable-css-custom-properties: false,
```
Comment (emyarod) [inline packages/components/src/globals/scss/_feature-flags.scss:35]: unsure of whether or not we want to toggle this here or just turn it on by default (maybe for v11?)

File: packages/themes/tasks/build.js [hunk 3]
```diff
@@ -254,6 +254,35 @@ function buildMixinsFile() {
...
+          consequent: t.BlockStatement(
+            Object.keys(tokens).flatMap(group => {
+              return tokens[group].flatMap(token => {
+                const name = formatTokenName(token);
+                return t.Assignment({
+                  id: t.Identifier(name),
+                  init: t.CallExpression({
+                    callee: t.Identifier('var'),
+                    arguments: [
+                      t.SassValue({
```
Comment (emyarod) [inline packages/themes/tasks/build.js:277]: I tried to see if there was a way to use `t.Identifier('custom-property-prefix')` somehow but I guess that's why `t.SassValue` was created in the first place (to handle this scenario)

**Comment by emyarod** [comment]
TODO: revise `@carbon/themes` test case for custom properties

**Review by asudoh — COMMENTED** [review]
LGTM 👍 at high-level, one thing I noticed running `gulp serve` is:

```
Message:
    src/globals/scss/vendor/@carbon/themes/scss/generated/_mixins.scss
Error: Undefined variable: "$custom-property-prefix".
        on line 105 of src/globals/scss/vendor/@carbon/themes/scss/generated/_mixins.scss, in mixin `carbon--theme`
        from line 36 of demo/scss/demo.scss
>> {$custom-property-prefix}-interactive-01,
```

**Comment by emyarod** [comment]
should be resolved now, added the missing import

File: packages/themes/__tests__/mixins-test.js [hunk 4]
```diff
@@ -31,6 +31,9 @@ describe('_mixins.scss', () => {
     const themeTests = Object.keys(themes).map(key => {
       const variable = `$carbon--theme--${key}`;
       const test = `
+        $feature-flags: (
```
Comment (emyarod) [inline packages/themes/__tests__/mixins-test.js]: it seems setting the feature flag does not affect the test result, need to look more into this

File: packages/themes/__tests__/mixins-test.js [hunk 5]
```diff
@@ -54,6 +57,10 @@ describe('_mixins.scss', () => {
     const { calls } = await render(`
       @import '../scss/themes';
 
+      $feature-flags: (
+        enable-css-custom-properties: false,
```
Comment (emyarod) [inline packages/themes/__tests__/mixins-test.js]: (same as above) it seems setting the feature flag does not affect the test result, need to look more into this

the commented lines below will cause the test to pass, but I think the issue is that the variables are becoming custom properties regardless of feature flag setting...

File: packages/themes/tasks/build.js (hunk 3 above)
Comment (joshblack) [reply packages/themes/tasks/build.js:277]: Yeah 😞 In the future we might be able to figure this out but for now I think `SassValue` is a useful shortcut until we're willing to invest more effort into this

File: packages/components/src/globals/scss/_feature-flags.scss (hunk 2 above)
Comment (joshblack) [reply packages/components/src/globals/scss/_feature-flags.scss:35]: This seems perfect 👍

**Review by joshblack — CHANGES_REQUESTED** [review]
Looks great, only comment was resolved in Slack, I think, where we were chatting about why `feature-flags` isn't getting applied. We might be able to update that part to:

```scss
global-variable-exists(feature-flags) and
map-get($feature-flags, 'enable-css-custom-properties') == true
```

And it should work 🙏

File: packages/components/demo/scss/demo.scss [hunk 6]
```diff
@@ -20,16 +20,19 @@
...
 //   )
 // );
+$feature-flags: (
+  enable-css-custom-properties: true,
+);
 
-@import './_theme.scss';
+@import '../../src/globals/scss/vendor/@carbon/themes/scss/mixins';
+@import '../../src/globals/scss/vendor/@carbon/themes/scss/generated/mixins';
```
Comment (joshblack) [inline packages/components/demo/scss/demo.scss]:
```suggestion
```

I believe `themes/scss/mixins` should be including `generated/mixins` by default 👍

File: packages/components/demo/scss/demo.scss [hunk 7]
```diff
@@ -20,16 +20,19 @@
...
 
 /// The theme name to use in the dev env.
 /// @type String
-$demo--carbon--theme-name: 'white' !default;
 
 /// The Carbon theme to use, chosen with `$demo--carbon--theme-name`.
 /// @type Map
-$carbon--theme: map-get($carbon--theme--list, $demo--carbon--theme-name);
+@include carbon--theme($carbon--theme--white);
```
Comment (joshblack) [inline packages/components/demo/scss/demo.scss]: Do we include carbon-component's globals for demo? If so, might not need to initialize here as it's done in `theme-tokens.scss` IIRC

**Review by joshblack — APPROVED** [review]
Looks great! Just had a couple comments around demo

File: packages/components/demo/scss/demo.scss (hunk 7 above)
Comment (emyarod) [reply packages/components/demo/scss/demo.scss]: yeah it looks like this is not needed although the theme imports might be coming from the elements side. we probably want to switch over to importing from `@carbon/themes` in v11

**Review by asudoh — APPROVED** [review]
Thanks @emyarod!

**Pull request merged**